src/kernel/MomentumBodyForceSrcElemKernel.C

```diff
@@ -0,0 +1,100 @@
+/*------------------------------------------------------------------------*/
+/*  Copyright 2014 Sandia Corporation.                                    */
+/*  This software is released under the license detailed                  */
+/*  in the file, LICENSE, which is located in the top-level Nalu          */
+/*  directory structure                                                   */
+/*------------------------------------------------------------------------*/
+
+#include "kernel/MomentumBodyForceSrcElemKernel.h"
+#include "AlgTraits.h"
+#include "master_element/MasterElement.h"
+#include "SolutionOptions.h"
+
+// template and scratch space
+#include "BuildTemplates.h"
+#include "ScratchViews.h"
+#include "utils/StkHelpers.h"
+
+// stk_mesh/base/fem
+#include <stk_mesh/base/Entity.hpp>
+#include <stk_mesh/base/MetaData.hpp>
+#include <stk_mesh/base/BulkData.hpp>
+#include <stk_mesh/base/Field.hpp>
+
+namespace sierra {
+namespace nalu {
+
+template <typename AlgTraits>
+MomentumBodyForceSrcElemKernel<AlgTraits>::MomentumBodyForceSrcElemKernel(
+  const stk::mesh::BulkData& bulkData,
+  const SolutionOptions& solnOpts,
+  ElemDataRequests& dataPreReqs)
+  : Kernel(),
+    ipNodeMap_(sierra::nalu::MasterElementRepo::get_volume_master_element(
+                 AlgTraits::topo_)
+                 ->ipNodeMap())
+{
+  const stk::mesh::MetaData& metaData = bulkData.mesh_meta_data();
+  densityNp1_ = get_field_ordinal(metaData, "density");
+
+  const std::vector<double>& solnOptsBodyForce =
+    solnOpts.get_bodyForce_vector(AlgTraits::nDim_);
+  for (int i = 0; i < AlgTraits::nDim_; i++)
+    bodyForce_(i) = solnOptsBodyForce[i];
+
+  MasterElement* meSCV =
+    sierra::nalu::MasterElementRepo::get_volume_master_element(
+      AlgTraits::topo_);
+  get_scv_shape_fn_data<AlgTraits>(
+    [&](double* ptr) { meSCV->shape_fcn(ptr); }, v_shape_function_);
+
+  // add master elements
+  dataPreReqs.add_cvfem_volume_me(meSCV);
+
+  // fields and data
+  dataPreReqs.add_gathered_nodal_field(densityNp1_, 1);
+  dataPreReqs.add_master_element_call(SCV_VOLUME, CURRENT_COORDINATES);
+}
+
+template <typename AlgTraits>
+MomentumBodyForceSrcElemKernel<AlgTraits>::~MomentumBodyForceSrcElemKernel()
+{
+}
+
+template <typename AlgTraits>
+void
+MomentumBodyForceSrcElemKernel<AlgTraits>::execute(
+  SharedMemView<DoubleType**>& /* lhs */,
+  SharedMemView<DoubleType*>& rhs,
+  ScratchViews<DoubleType>& scratchViews)
+{
+  SharedMemView<DoubleType*>& v_densityNp1 =
+    scratchViews.get_scratch_view_1D(densityNp1_);
+  SharedMemView<DoubleType*>& v_scv_volume =
+    scratchViews.get_me_views(CURRENT_COORDINATES).scv_volume;
+
+  for (int ip = 0; ip < AlgTraits::numScvIp_; ++ip) {
+    const int nearestNode = ipNodeMap_[ip];
+    DoubleType densityIp = 0.0;
+
+    for (int ic = 0; ic < AlgTraits::nodesPerElement_; ++ic) {
+      const DoubleType r = v_shape_function_(ip, ic);
+      densityIp += r * v_densityNp1(ic);
+    }
+
```

Comment:
@rcknaus sorry, not sure I follow. You mean renaming the class? What name do you have in mind?

Reply:
Yeah, and the input file name.  Maybe "body acceleration" or something.  Just some word to distinguish `rho f_specified` vs `F_specified`

Reply:
Ok got it. Do you think it would make more sense to actually change the Elem version to be consistent with the Node version? I could remove `density` from the Elem version.

Reply:
Both are good options.   Depends what you wanted for the body force here.

Reply:
I made them consistent. Thanks for pointing that out